cmd/snapd-generator/main.c

```diff
@@ -97,3 +90,80 @@ int main(int argc, char **argv)
 	fprintf(f, "WantedBy=local-fs.target\n");
 	return 0;
 }
+
+int ensure_fusesquashfs_inside_container(const char *normal_dir)
+{
+	// check if we are running inside a container
+	if (system("systemd-detect-virt --container --quiet") != 0) {
+		return 0;
+	}
+
+	char *fstype = NULL;
+	if (system("which squashfuse > /dev/null") == 0) {
+		fstype = "fuse.squashfuse";
+	} else if (system("which snapfuse > /dev/null") == 0) {
+		fstype = "fuse.snapfuse";
+	} else {
+		fprintf(stderr,
+			"cannot find squashfuse or snapfuse executable\n");
+		return 2;
+	}
+
+	DIR *units_dir SC_CLEANUP(sc_cleanup_closedir) = NULL;
+    units_dir = opendir("/etc/systemd/system");
+    if (units_dir == NULL) {
+		fprintf(stderr,
+			"cannot open /etc/systemd/system directory: %m\n");
+        return 2;
+    }
+
+	char fname[PATH_MAX + 1] = { 0 };
+
+    struct dirent *ent;
+    while (ent = readdir(units_dir)) {
+		if (!(sc_startswith(ent->d_name, "snap-") && sc_endswith(ent->d_name, ".mount"))) {
+			continue;
+		}
+		sc_must_snprintf(fname, sizeof fname,
+			"%s/%s.d", normal_dir, ent->d_name);
+		if (mkdir(fname, 0755) != 0) {
+			// generators are run over clean directory, no need to handle EEXIST
```

Comment:
Is this actually true? The generator runs a lot, everytime "systemctl daemon-reload" is called

Reply:
Yes, daemon-reload cleans run/systemd/generator directory, at least this is what I saw every time on 18.04.

Reply:
Thank you :)

Reply:
the generator receives an empty directory. However, some other generator that is running in parallel may too create .d drop in for the same unit.......... so you might need to handle EEXIST